### ISSUE TITLE
TURN Server NOT Working, Broswer still prompt: Error Code: 1002

### ISSUE THREAD

**Issue by congyijun** [bigbluebutton/bigbluebutton#9580]
**Describe the bug**
I refer to BigBlueButton installation guideline setup BBB and GreenLight, and also verify the service working fine. But just found corporate computer (win10) could NOT join conference call by audio. My corporate computer access Internet by McAfee proxy server, and behind firewall - CheckPoint. 

I already verified or can say confirmed, the BBB and Greenlight working stable if the client from Internet by 4G or home Internet. So I follow the TURN Server guideline, setup additional Ubuntun 18.04, just support TURN Server. 

By the way,  the BBB and Greenlight host in one Ubuntu 16.04. Please refer to  turn-stun-servers.xml on BBB and Greenlight server

-----------------------------
<?xml version="1.0" encoding="UTF-8"?>
<beans xmlns="http://www.springframework.org/schema/beans"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.springframework.org/schema/beans
        http://www.springframework.org/schema/beans/spring-beans-2.5.xsd">

    <bean id="stun0" class="org.bigbluebutton.web.services.turn.StunServer">
        <constructor-arg index="0" value="stun:turn.bbb.public.domain"/>
    </bean>

    <bean id="turn0" class="org.bigbluebutton.web.services.turn.TurnServer">
        <constructor-arg index="0" value="wo0hMq---key"/>
        <constructor-arg index="1" value="turns:turn.bbb.public.domain:443?transport=tcp"/>
        <constructor-arg index="2" value="86400"/>
    </bean>

    <bean id="stunTurnService"
            class="org.bigbluebutton.web.services.turn.StunTurnService">
        <property name="stunServers">
            <set>
                <ref bean="stun0"/>
            </set>
        </property>
        <property name="turnServers">
            <set>
                <ref bean="turn0"/>
            </set>
        </property>
    </bean>
</beans>
---------------------------------------

I also try command "tcpdump port 443" on TURN server, but didn't capture any traffic from windows 10. Just worry the BBB main server didn't pass the TURN server information to windows 10 (client)

By the way, i also refer to BigBlueButton guideline configure BBB server. Because it behind firewall.

**To Reproduce**
1. Join the conference call
2. Click the "Microphone"
3. Click the "Allow access the Microphone"
4. See Error Code: 1002

**Expected behavior**
By firefox, Windows10 could join audio and video conference call successful

**Actual behavior**
By firefox,  Windoes 10 could NOT join audio and video conference call

**Screenshots**
![08](https://user-images.githubusercontent.com/62802149/82155039-b8296300-98a4-11ea-8148-97704c30ef2e.png)

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Firefox
 - 68.7

**Additional context**
If request additional information support troubleshooting, please feel free to let me know.

Many thanks!

**Comment by ichdasich** [comment]
Heho,
this is a bug in firefox, which does not handle WebRTC (and hence TURN) via the proxy api. Hence, you can not contact the TURN server. See the following two bugs:

https://bugzilla.mozilla.org/show_bug.cgi?id=1505329
https://bugzilla.mozilla.org/show_bug.cgi?id=1239006

tldr: With the next ESR based on FF78, this should be fixed.

**Comment by congyijun** [comment]
> 
> 
> Heho,
> this is a bug in firefox, which does not handle WebRTC (and hence TURN) via the proxy api. Hence, you can not contact the TURN server. See the following two bugs:
> 
> https://bugzilla.mozilla.org/show_bug.cgi?id=1505329
> https://bugzilla.mozilla.org/show_bug.cgi?id=1239006
> 
> tldr: With the next ESR based on FF78, this should be fixed.

Hello ichdasich

If there is bug impact firefox couldn;t connect to TURN server, can I try the Chrome or Microsoft Edge?  Still behind proxy server and firewall.

**Comment by ichdasich** [comment]
I would give chrome/edge a try. Please report back, as I would be interested in the results as well.

**Comment by congyijun** [comment]
> 
> 
> I would give chrome/edge a try. Please report back, as I would be interested in the results as well.

Hello ichdasich

Bad news,  Edge reminder Browser reminder NOT support. And Chrome still prompt Error code : 1002.    Feel Discouraged

Actually,  i don't have any idea find the issue. And i total refer to configuration guideline, setup TURN server step by step/

**Comment by ichdasich** [comment]
Well, the issue is the use of the web-proxy without routing. Turn does apparently not really help then. Can you try a most recent firefox?

**Comment by congyijun** [comment]
> 
> 
> Well, the issue is the use of the web-proxy without routing. Turn does apparently not really help then. Can you try a most recent firefox?

From my understanding, the BBB server will provide TURN server list to windows computer. So the windows computer could connect to TURN server (https) by proxy server.

I just try connect to https://test.bigbluebutton.org/, also get the same error code: 1002. my computer also behind proxy and firewall.

Tomorrow, i can try the last firefox. Keep you posted 

Many thanks!

**Comment by congyijun** [comment]
@ichdasich 

Let me share some strange thing with you. I just try connect to BBB server again in my corporate network, also behind firewall and proxy server. 

I also enable F12 on Firefox, review the connection log for troubleshooting purpose. I can see the BBB server pass the STUN and TURN server url to Firefox. And then the i  found the Firefox try connect to BBB server by WS, mean websocket.    

My understanding the websocket connection failure result in audio conference call couldn;t be connected

I am not sure the following url connection correct or not, because i already setup TURN server and associate with my BBB server. 

https://public domain/ws

------------------
/etc/bigbluebutton/nginx/sip.nginx

location /ws {
        proxy_pass https://public domain:7443;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_read_timeout 6h;
        proxy_send_timeout 6h;
        client_body_timeout 6h;
        send_timeout 6h;
--------------------------

please refer to following screenshot.   TOTALLY, NO IDEA

![08](https://user-images.githubusercontent.com/62802149/82337657-9ce26300-9a1e-11ea-8349-fa88959bae18.png)
![09](https://user-images.githubusercontent.com/62802149/82337663-9e139000-9a1e-11ea-8a4c-4cdfd3e8f634.png)
![10](https://user-images.githubusercontent.com/62802149/82337667-9f44bd00-9a1e-11ea-8870-364658dfeba6.png)
![11](https://user-images.githubusercontent.com/62802149/82337673-a075ea00-9a1e-11ea-971a-62ef29d0c5eb.png)

**Comment by ichdasich** [comment]
So... you access those resources via the proxy at 10.50.X.X:8080 in your RFC1918 network?

Then these are pretty much just the associated web requests; WebRTC in the background still fails, according to the two mozilla bugs I posted above.

**Comment by congyijun** [comment]
@ichdasich 

Correct, the 10.50.XXX.XXX:8080, as proxy server in our corporate network. i would try last one, download last firefox and connect to our corporate network, try it again

**Comment by basisbit** [comment]
the described problem usually happens when a company is using some proxy server software which does not support WebSocket. Sadly, these still exist even today.
To test your proxy if it supports WebSocket, use https://www.websocket.org/echo.html and just use the default address.
If that works, but you still can't connect to your server, please use https://test.webrtc.org/ to test your WebRTC environment. We'd need a screenshot of these two websites results.

**Comment by congyijun** [comment]
@basisbit 

Many thanks for your comments. Please refer to following screenshot. By the way, i already followed BBB website official document, setup dedicate TURN server. And i also could get the TURN server URL in Firefox
![05](https://user-images.githubusercontent.com/62802149/84056279-95442780-a9e8-11ea-8c39-8281ea737abd.PNG)

![06](https://user-images.githubusercontent.com/62802149/84056794-4f3b9380-a9e9-11ea-9757-eae8564bfdb8.PNG)

![07](https://user-images.githubusercontent.com/62802149/84056816-56fb3800-a9e9-11ea-9db6-71ea9fd93267.PNG)

**Comment by GhaziTriki** [comment]
Closing as does seem to be related with user network restrictions. Please re-open if needed.
